File: src/main/java/com/ioannuwu/inline/ui/render/elements/graphiccomponents/TextComponent.java
package com.ioannuwu.inline.ui.render.elements.graphiccomponents;

import com.intellij.openapi.editor.Inlay;
import com.intellij.openapi.editor.markup.TextAttributes;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;

public interface TextComponent extends GraphicsComponent {

    class Base implements TextComponent {

        protected final FontData fontData;

        protected final Color color;
        protected final String text;

        public Base(@NotNull FontData fontData, @NotNull Color color, @NotNull String text) {
            this.fontData = fontData;
            this.color = color;
            this.text = text;
        }

        @Override
        public void draw(@NotNull Inlay inlay, @NotNull Graphics g, @NotNull Rectangle targetRegion,
                         @NotNull TextAttributes textAttributes) {

            g.setFont(fontData.font());
            g.setColor(color);

            double centerX = width() / 2f + targetRegion.x;

            var fontMetrics = fontData.fontMetrics();

            g.drawString(text,
                    (int) (centerX - fontMetrics.stringWidth(text) / 2) + fontMetrics.charWidth('a'),
                    targetRegion.y + fontData.lineHeight() * 3 / 4);
        }

        @Override
        public int width() {
            var fontMetrics = fontData.fontMetrics();
            return fontMetrics.stringWidth(text) + fontMetrics.charWidth('a') * 2;
        }
    }

    class WithDotAtTheEnd extends Base {

        public WithDotAtTheEnd(@NotNull FontData fontData, @NotNull Color color, @NotNull String text) {
            super(fontData, color, text + ".");
        }
    }
}

File: src/main/kotlin/com/ioannuwu/inline/graphics/TextComponent.kt
package com.ioannuwu.inline.graphics

import java.awt.Color
import java.awt.Graphics
import java.awt.Rectangle

interface TextComponent : GraphicsComponentKt {

    override fun drawFancy(g: Graphics, targetRegion: Rectangle, color: Color)

    class AfterLineText(
        private val fontData: FontData,
        private val textColor: Color,
        private val text: String,
    ) : TextComponent {

        override fun drawFancy(g: Graphics, targetRegion: Rectangle, color: Color) {
            g.color = color
            g.font = fontData.font()

            val centerX = width / 2f + targetRegion.x
            val fontMetrics = fontData.fontMetrics()

            g.drawString(text,
                (centerX - fontMetrics.stringWidth(text).toFloat() / 2 + fontMetrics.charWidth('a')).toInt(),
                targetRegion.y + fontData.lineHeight() * 3 / 4)
        }

        override fun draw(g: Graphics, targetRegion: Rectangle) {
            drawFancy(g, targetRegion, textColor)
        }

        override val width: Int
            get() {
                val fontMetrics = fontData.fontMetrics()
                return fontMetrics.stringWidth(text) + fontMetrics.charWidth('a') * 2
            }
    }


}